### PR TITLE
Gameengine

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: implementation/commontest/src/main/java/org/inaetics/dronessimulator/test/concurrent/ConcurrentExecute.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ConcurrentExecute {
```
Comment (jjkester) [inline implementation/commontest/src/main/java/org/inaetics/dronessimulator/test/concurrent/ConcurrentExecute.java]: Maybe document what this class is for exactly.

File: implementation/commontest/src/main/java/org/inaetics/dronessimulator/test/concurrent/ConcurrentJobEntry.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
+package org.inaetics.dronessimulator.test.concurrent;
+
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class ConcurrentJobEntry {
```
Comment (jjkester) [inline implementation/commontest/src/main/java/org/inaetics/dronessimulator/test/concurrent/ConcurrentJobEntry.java]: Maybe document what this class is for

File: implementation/commontest/src/main/java/org/inaetics/dronessimulator/test/concurrent/TimeoutHandler.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
+package org.inaetics.dronessimulator.test.concurrent;
+
+import java.util.concurrent.PriorityBlockingQueue;
+
+public class TimeoutHandler extends Thread {
```
Comment (jjkester) [inline implementation/commontest/src/main/java/org/inaetics/dronessimulator/test/concurrent/TimeoutHandler.java]: Maybe document what this class is for

File: implementation/discovery/etcd/src/test/java/org/inaetics/dronessimulator/discovery/etcd/EtcdDiscovererTest.java [hunk 4]
```diff
@@ -10,7 +10,8 @@
 import java.util.HashMap;
 import java.util.Map;
 
-import static org.junit.Assert.*;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
```
Comment (jjkester) [inline implementation/discovery/etcd/src/test/java/org/inaetics/dronessimulator/discovery/etcd/EtcdDiscovererTest.java:14]: Are these changes intentional or a side effect of some cleanup magic?

File: implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/gameevent/DestroyHealthEntityEvent.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
+package org.inaetics.dronessimulator.gameengine.common.gameevent;
+
+
```
Comment (jjkester) [inline implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/gameevent/DestroyHealthEntityEvent.java]: Extra newline

File: implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/state/HealthGameEntity.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package org.inaetics.dronessimulator.gameengine.common.state;
+
+import lombok.EqualsAndHashCode;
+import org.inaetics.dronessimulator.common.D3Vector;
+import org.inaetics.dronessimulator.common.protocol.EntityType;
+
+@EqualsAndHashCode(callSuper=true)
+public abstract class HealthGameEntity<C extends GameEntity<C>> extends GameEntity<C> {
```
Comment (jjkester) [inline implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/state/HealthGameEntity.java:8]: Does this compile?

File: implementation/gameengine/identifier-mapper/src/main/java/org/inaetics/dronessimulator/gameengine/identifiermapper/IIdentifierMapper.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
+package org.inaetics.dronessimulator.gameengine.identifiermapper;
+
+public interface IIdentifierMapper {
```
Comment (jjkester) [inline implementation/gameengine/identifier-mapper/src/main/java/org/inaetics/dronessimulator/gameengine/identifiermapper/IIdentifierMapper.java]: Pretty unreadable class name...

File: implementation/commontest/pom.xml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+        <artifactId>dronessimulator</artifactId>
+        <groupId>org.inaetics</groupId>
+        <version>0.1</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <packaging>jar</packaging>
+
+    <groupId>org.inaetics.dronessimulator</groupId>
+    <artifactId>common-test</artifactId>
```
Comment (jjkester) [inline implementation/commontest/pom.xml]: Why not put this in/under `common`?

File: implementation/pom.xml [hunk 9]
```diff
@@ -15,49 +15,11 @@
         <module>discovery</module>
         <module>visualisation</module>
         <module>gameengine</module>
+        <module>drone</module>
+        <module>instancemanager</module>
+        <module>commontest</module>
```
Comment (jjkester) [inline implementation/pom.xml]: Does not match artifactId of package (which has a dash).

**Review by jjkester — CHANGES_REQUESTED** [review]
In general this stuff *really* needs some documentation since I was not always able to figure out what some things are supposed to do.

File: implementation/discovery/etcd/src/test/java/org/inaetics/dronessimulator/discovery/etcd/EtcdDiscovererTest.java (hunk 4 above)
Comment (ghost) [reply implementation/discovery/etcd/src/test/java/org/inaetics/dronessimulator/discovery/etcd/EtcdDiscovererTest.java:14]: Intentional primary effect of optimizing the imports in each file.

File: implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/state/HealthGameEntity.java (hunk 6 above)
Comment (ghost) [reply implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/state/HealthGameEntity.java:8]: Yep. C is the type of the child. This generic is usually used in places where you want to define a function that returns the specific type which is inheriting. In this case, it is necassary for the function deepcopy which should return something of the C child.

File: implementation/commontest/pom.xml (hunk 8 above)
Comment (ghost) [reply implementation/commontest/pom.xml]: common is not needed by all packages and I would rather not include common where common-test would suffice.

**Comment by ghost** [comment]
Meer docu voor de gameengine komt na de milestone

File: implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/state/HealthGameEntity.java (hunk 6 above)
Comment (jjkester) [reply implementation/gameengine/common/src/main/java/org/inaetics/dronessimulator/gameengine/common/state/HealthGameEntity.java:8]: Ah, it kind of looks strange but well, if it works :)

**Review by jjkester — APPROVED** [review]
